corehq/apps/sms/static/sms/js/phone_number_filter.js

```diff
@@ -1,42 +1,57 @@
-
-    this.phone_number_options = ko.pureComputed(function () {
-        if (model.contact_type() === 'cases') {
-            return [PHONE_NUMBER_SELECT_OPTIONS[0]];
-        }
-        return PHONE_NUMBER_SELECT_OPTIONS;
-    });
+hqDefine("sms/js/phone_number_filter", function() {
+    function SMSPhoneNumberFilterViewModel(initial_value, groups) {
+        var PHONE_NUMBER_SELECT_OPTIONS = [
+                {id: "has_phone_number", text: gettext("That have phone numbers")},
+                {id: "no_phone_number", text: gettext("That do not have phone numbers")},
+            ],
+            model = this;
 
-    this.show_phone_filter = ko.pureComputed(function () {
-        return model.filter_type() === 'phone_number';
-    });
+        this.filter_type = ko.observable(initial_value.filter_type || 'phone_number');
+        this.phone_number_filter = ko.observable(initial_value.phone_number_filter);
+        this.has_phone_number = ko.observable(initial_value.has_phone_number);
+        this.contact_type = ko.observable(initial_value.contact_type);
+        this.groups = ko.observableArray(groups);
+        this.selected_group = ko.observable(initial_value.selected_group);
+        this.verification_status = ko.observable(initial_value.verification_status);
 
-    this.show_contact_filter = ko.pureComputed(function () {
-        return model.filter_type() === 'contact';
-    });
+        this.phone_number_options = ko.pureComputed(function () {
+            if (model.contact_type() === 'cases') {
+                return [PHONE_NUMBER_SELECT_OPTIONS[0]];
+            }
+            return PHONE_NUMBER_SELECT_OPTIONS;
+        });
 
-    this.show_group_filter = ko.pureComputed(function () {
-        return model.show_contact_filter() && model.contact_type() === 'users';
-    });
+        this.show_phone_filter = ko.pureComputed(function () {
+            return model.filter_type() === 'phone_number';
+        });
 
-    this.can_edit_has_phone_number = ko.pureComputed(function () {
-        return model.show_contact_filter() && model.contact_type() === 'cases';
-    });
+        this.show_contact_filter = ko.pureComputed(function () {
+            return model.filter_type() === 'contact';
+        });
+
+        this.show_group_filter = ko.pureComputed(function () {
+            return model.show_contact_filter() && model.contact_type() === 'users';
+        });
+
+        this.can_edit_has_phone_number = ko.pureComputed(function () {
+            return model.show_contact_filter() && model.contact_type() === 'cases';
+        });
+
+        this.show_verification_filter = ko.pureComputed(function () {
+            return model.show_contact_filter() && model.has_phone_number() === 'has_phone_number';
+        });
+    }
 
-    this.show_verification_filter = ko.pureComputed(function () {
-        return model.show_contact_filter() && model.has_phone_number() === 'has_phone_number';
+    $(function() {
```

Comment:
If this filter is used more than once on a page, this will attempt to re-initialize filters, since this script will be included once per filter and each time will grab all `.report-filter-phone-number` elements. It shouldn't actually break anything because this filter is only used in one place, but it isn't a great pattern. What would you think of doing this in https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/static/reports/js/filters.js ?

Reply:
Not entirely sure how I feel about this pattern, since it means all of these jquery requests will run on every reports page, but I'm also not sure how to improve it.

Reply:
"these jquery requests" => meaning the checks for the various widget classes? I'm not especially worried (getting elements by class name is fast), but I'm also not that attached to this pattern, happy to iterate on it as we keep migrating.